### PR TITLE
Don't make statement nonterminals match pattern nonterminals

### PR DESCRIPTION
**Pull request by compiler-errors**
Right now, the heuristic we use to check if a token may begin a pattern nonterminal falls back to `may_be_ident`:
https://github.com/rust-lang/rust/blob/ef71f1047e04438181d7cb925a833e2ada6ab390/compiler/rustc_parse/src/parser/nonterminal.rs#L21-L37

This has the unfortunate side effect that a `stmt` nonterminal eagerly matches against a `pat` nonterminal, leading to a parse error:
```rust
macro_rules! m {
    ($pat:pat) => {};
    ($stmt:stmt) => {};
}

macro_rules! m2 {
    ($stmt:stmt) => {
        m! { $stmt }
    };
}

m2! { let x = 1 }
```

This PR fixes it by more accurately reflecting the set of nonterminals that may begin a pattern nonterminal.

As a side-effect, I modified `Token::can_begin_pattern` to work correctly and used that in `Parser::nonterminal_may_begin_with`.

### CODE REVIEW

File: compiler/rustc_ast/src/token.rs [hunk 1]
```diff
@@ -475,24 +478,34 @@ impl Token {
...
-            // DotDotDot is no longer supported
-            | DotDot | DotDotDot | DotDotEq      // ranges
-            | Lt | BinOp(Shl)                    // associated path
-            | ModSep                    => true, // global path
-            Interpolated(ref nt) => matches!(&nt.0, NtLiteral(..) |
-                NtPat(..)     |
-                NtBlock(..)   |
-                NtPath(..)),
+    /// **NB**: Take care when modifying this function, since it will change
+    /// the stable set of tokens that are allowed to match an pat nonterminal.
```
Comment (nnethercote) [inline compiler/rustc_ast/src/token.rs:482]:
```suggestion
    /// the stable set of tokens that are allowed to match a pat nonterminal.
```

File: compiler/rustc_ast/src/token.rs [hunk 2]
```diff
@@ -475,24 +478,34 @@ impl Token {
...
+        match &self.uninterpolate().kind {
+            Ident(..) |                          // box, ref, mut, and other identifiers (can stricten)
+            OpenDelim(Delimiter::Parenthesis) |  // tuple pattern
+            OpenDelim(Delimiter::Bracket) |      // slice pattern
+            BinOp(And) |                  // reference
+            BinOp(Minus) |                // negative literal
+            AndAnd |                             // double reference
+            Literal(_) |                        // literal
+            DotDot |                             // range pattern (future compat)
+            DotDotDot |                          // range pattern (future compat)
```
Comment (nnethercote) [inline compiler/rustc_ast/src/token.rs:493]: AFAICT, the net effect here:
- All identifiers now match, i.e. more than before
- `DotDotEq` no longer matches
- `BinOp(Or)` matches if `allow_leading_or` is true
- `NtExpr`, `NtMeta`, `NtExpr` now match

Is that right?

File: compiler/rustc_parse/src/parser/nonterminal.rs [hunk 3]
```diff
@@ -66,26 +66,8 @@ impl<'a> Parser<'a> {
...
-                token::Lt |                                 // path (UFCS constant)
-                token::BinOp(token::Shl) => true,           // path (double UFCS)
-                // leading vert `|` or-pattern
-                token::BinOp(token::Or) => matches!(kind, NonterminalKind::PatWithOr),
-                token::Interpolated(nt) => may_be_ident(&nt.0),
-                _ => false,
-            }
-            }
+            NonterminalKind::PatParam { .. } => token.can_begin_pattern(false),
+            NonterminalKind::PatWithOr => token.can_begin_pattern(true),
```
Comment (nnethercote) [inline compiler/rustc_parse/src/parser/nonterminal.rs:70]: And the net effect here is that `NtStmt` no longer matches. Which I think is the change that actually fixes the original problem, right?

**Comment by nnethercote** [comment]
How did you find this problem? Just curious.

This appears to change more than is necessary to fix the cited bug. Is that intended?

File: compiler/rustc_ast/src/token.rs (hunk 2 above)
Comment (compiler-errors) [reply compiler/rustc_ast/src/token.rs:493]: Yes, though it should be noted that can_begin_pattern was previously only used for diagnostics and was previously just a rip off of can_begin_expr from above.

A better way to look at the changes here is to compare it to the previous match that was located in nonterminal_may_begin_with. This PR copies that logic over, adds an uninterpolate call for consistency with can_begin_expe (rather than manually handling the NtIdent arm), and adjusting the match arm for the Interpolated token.

To sum it up, I believe the only token kind that now no longer matches pat nonterminals is the statement nonterminal.

File: compiler/rustc_parse/src/parser/nonterminal.rs (hunk 3 above)
Comment (compiler-errors) [reply compiler/rustc_parse/src/parser/nonterminal.rs:70]: Yes

**Comment by compiler-errors** [comment]
> How did you find this problem? Just curious.

I noticed it while authoring #120218, and thinking a bit about why there was a `Token::can_begin_pattern` (that, for the record, I believe I added for diagnostics purposes originally) that was not being used by `Parser::nonterminal_may_begin_with`.

> This appears to change more than is necessary to fix the cited bug. Is that intended?

Yes. I don't see why we need to maintain an incorrect `Token::can_begin_pattern` just for diagnostics rather than just fixing it to actually reflect the set of tokens that can match `pat` nonterminals and using it everywhere.

**Comment by nnethercote** [comment]
Thanks for the explanations. r=me with the "a"/"an" nit fixed.

**Comment by compiler-errors** [comment]
Actually, given that this is a change to a contract (the macro matchers), I'll pass this by t-lang as well.

@rustbot label: +I-lang-nominated

**Comment by nikomatsakis** [comment]
@rfcbot reviewed

**Comment by tmandry** [comment]
@rfcbot reviewed

File: compiler/rustc_parse/src/parser/path.rs [hunk 4]
```diff
@@ -381,7 +381,10 @@ impl<'a> Parser<'a> {
                     if self.may_recover()
                         && prev_token_before_parsing.kind == token::PathSep
                         && (style == PathStyle::Expr && self.token.can_begin_expr()
-                            || style == PathStyle::Pat && self.token.can_begin_pattern())
+                            || style == PathStyle::Pat
```
Comment (compiler-errors) [inline compiler/rustc_parse/src/parser/path.rs:384]: This is uh... weird but also I have no idea what this recovery is doing, so I'm going to ignore it.

File: compiler/rustc_ast/src/token.rs [hunk 5]
```diff
@@ -503,34 +506,46 @@ impl Token {
             PathSep                            | // global path
             Lifetime(..)                      | // labeled loop
             Pound                             => true, // expression attributes
-            Interpolated(ref nt) => matches!(&**nt, NtLiteral(..) |
-                NtExpr(..)    |
-                NtBlock(..)   |
-                NtPath(..)),
+            Interpolated(ref nt) =>
```
Comment (compiler-errors) [inline compiler/rustc_ast/src/token.rs:509]: I sorted and formatted this
